### PR TITLE
Avoid compiler-generated writeThis() methods for array classes

### LINKED ISSUE #14467: Printing first-class functions and using scans on BlockDist don't mix
### Summary of Problem

When a program tries to both (1) print / stringify a first-class function and (2) compute a scan, problems occur due to arrays of sync variables used to implement the scan and a speculative attempt to resolve what writing those arrays out would do.

Specifically, I believe what's happening is that the compiler's dynamic dispatch machinery starts analyzing FCF->writeThis() which causes writeThis() to be considered / resolved for all classes / subclasses of object (or at least, more than just the first-class function).  The scan code contains an array of `sync bool` elements, and `sync`s don't support being written by default.  So if the scan code has been resolved, the compiler-generated writeThis() for the class implementing the array of bools is resolved as part of this dynamic dispatch processing, and we hit the error (even though no calls to write or stringify the array of sync bools are made).

I'm not sure how best to address this, but some thoughts I have offhand are:
* are we working too hard to resolve FCF->writeThis()?  Do we really need to be looking across so much of the object hierarchy for this case?  And/or could we save compile-time by not doing so?
* if we must do this resolution, it seems that we ought to be more resilient to resolution errors like this somehow—perhaps by ignoring them, or turning them into execution-time halts (?), particularly for compiler-generated routines?
* or maybe we could avoid creating compiler-generated writeThis() routines for types for which it would only be an error and have the error be manifested in another way if someone did try to write such a type out?

### Steps to Reproduce

**Source Code:**

```chapel
use BlockDist;

var A = newBlockArr({1..10}, real);
var B = + scan A;

proc foo() {
  writeln("In foo");
  return 42;
}

writeln("foo is: " + foo:string);
```

**Compile command:**
`chpl foo.chpl`

**Compiler output:**

(because I suspect that nobody will correlate this failure mode to this issue without it)

```
$CHPL_HOME/modules/standard/IO.chpl:3157: error: sync variables cannot currently be written - apply readFE/readFF() to those variables first
while processing the following Chapel call chain:
  $CHPL_HOME/modules/internal/ChapelSyncvar.chpl:178: _syncvar.writeThis(x) [internal module]
  $CHPL_HOME/modules/standard/IO.chpl:3157: _write_one_internal(_channel_internal: qio_channel_ptr_t, param kind: iokind, x, loc: borrowed locale?)
  $CHPL_HOME/modules/standard/IO.chpl:3201: channel.writeIt(x)
  $CHPL_HOME/modules/standard/IO.chpl:3235: <~>(ch: channel, x)
  $CHPL_HOME/modules/internal/DefaultRectangular.chpl:1616: recursiveArrayWriter(idx: 1*int(64), dim: int(64), last: bool) [internal module]
  $CHPL_HOME/modules/internal/DefaultRectangular.chpl:1774: chpl_serialReadWriteRectangularHelper(f, arr, dom) [internal module]
  $CHPL_HOME/modules/internal/DefaultRectangular.chpl:1575: chpl_serialReadWriteRectangular(f, arr, dom) [internal module]
  $CHPL_HOME/modules/internal/DefaultRectangular.chpl:1562: chpl_serialReadWriteRectangular(f, arr) [internal module]
  $CHPL_HOME/modules/internal/DefaultRectangular.chpl:1552: DefaultRectangularArr.dsiSerialReadWrite(f) [internal module]
  $CHPL_HOME/modules/internal/DefaultRectangular.chpl:1779: DefaultRectangularArr.dsiSerialWrite(f) [internal module]
  $CHPL_HOME/modules/internal/ChapelArray.chpl:2876: [].writeThis(f) [internal module]
  $CHPL_HOME/modules/standard/IO.chpl:3157: _write_one_internal(_channel_internal: qio_channel_ptr_t, param kind: iokind, x, loc: borrowed locale?)
  $CHPL_HOME/modules/standard/IO.chpl:3201: channel.writeIt(x)
  $CHPL_HOME/modules/standard/IO.chpl:3215: channel.readwrite(x)
  $CHPL_HOME/modules/internal/ChapelIO.chpl:281: writeThisFieldsDefaultImpl(writer, x, first: bool) [internal module]
  $CHPL_HOME/modules/internal/ChapelIO.chpl:337: writeThisDefaultImpl(writer, x) [internal module]
  $CHPL_HOME/modules/dists/ReplicatedDist.chpl:0: LocReplicatedArr.writeThis(f) [compiler-generated]
  $CHPL_HOME/modules/standard/IO.chpl:3151: _write_one_internal(_channel_internal: qio_channel_ptr_t, param kind: iokind, x, loc: borrowed locale?)
  $CHPL_HOME/modules/standard/IO.chpl:3201: channel.writeIt(x)
  $CHPL_HOME/modules/standard/IO.chpl:3235: <~>(ch: channel, x)
  $CHPL_HOME/modules/internal/DefaultAssociative.chpl:798: DefaultAssociativeArr.dsiSerialReadWrite(f) [internal module]
  $CHPL_HOME/modules/internal/DefaultAssociative.chpl:848: DefaultAssociativeArr.dsiSerialWrite(f) [internal module]
  $CHPL_HOME/modules/internal/ChapelArray.chpl:2876: [].writeThis(f) [internal module]
  $CHPL_HOME/modules/standard/IO.chpl:3157: _write_one_internal(_channel_internal: qio_channel_ptr_t, param kind: iokind, x, loc: borrowed locale?)
  $CHPL_HOME/modules/standard/IO.chpl:3201: channel.writeIt(x)
  $CHPL_HOME/modules/standard/IO.chpl:3215: channel.readwrite(x)
  $CHPL_HOME/modules/internal/ChapelIO.chpl:281: writeThisFieldsDefaultImpl(writer, x, first: bool) [internal module]
  $CHPL_HOME/modules/internal/ChapelIO.chpl:337: writeThisDefaultImpl(writer, x) [internal module]
  $CHPL_HOME/modules/dists/ReplicatedDist.chpl:0: ReplicatedArr.writeThis(f) [compiler-generated]
```

### Configuration Information

- Output of `chpl --version`: 1.20

### PR DESCRIPTION
**Pull request by bradcray**
This PR avoids having the compiler generate writeThis() methods for classes that implement arrays because Chapel arrays are printed out using dsi routines, so having the compiler generate these writeThis() methods is wasted work and generated code (and as a result, I'm hoping will reduce compile-time slightly).

The approach taken here is in two parts:  (1) we don't generate writeThis() routines for array classes that inherit from AbsBaseArr;  (2) for the "local array" classes of distributions, we add an explicit writeThis() routine that essentially just halt()s if it's ever called.

My original motivation for this PR is issue #14467 which ran into problems because distributed arrays of sync variables would get compiler-generated `writeThis()` overloads for their local array classes which would be resolved during virtual method resolution, hitting the compilerError() in the sync variable writeThis() code.  The only thing that's required to close that issue is the writeThis() overload on the LocReplicatedArr class, but a similar issue could occur if someone created any other distributed array of syncs or singles.  So here, I've taken the approach of adding writeThis() overloads to all local array classes.  Having the compiler not insert them for the global array classes is both a safeguard for non-distributed arrays, and also an attempt to avoid creating unused writeThis() calls that (I expect) result in a lot of generated code.

I think the better way to solve #14467 is probably to bite the bullet and come up with a definition of how sync variables should write/read themselves since they are somewhat of an outlier in this regard.  But even if we did that, avoiding compiler-generate writeThis() routines that will never be used still seems like it could have value.

Resolves #14467.

### CODE REVIEW

**Comment by bradcray** [comment]
Based on some simple tests of hello/hello4 on my laptop, there's not as much difference in terms of compile-time or generated code size (for the latter, I assume this suggests that we're more aggressive about pruning out unused methods than I would've guessed?).  The AST size is reduced modestly (in the 100's of nodes), but nothing to brag about.

File: compiler/passes/buildDefaultFunctions.cpp [hunk 1]
```diff
@@ -1647,6 +1647,25 @@ static bool inheritsFromError(Type* t) {
 }
 
 
+static bool inheritsFromAbsBaseArr(Type* t) {
```
Comment (mppf) [inline compiler/passes/buildDefaultFunctions.cpp:1650]: Isn't this the same as `isArrayImplType` ?

Comment (bradcray) [reply compiler/passes/buildDefaultFunctions.cpp:1650]: Good catch, thanks!

**Comment by bradcray** [comment]
@ronawho: Not sure whether you've been tracking this PR, but it should resolve the variations on #14667 that you were running into if you were interested in backing out the other workaround (if you took it).

**Pull request merged**